cmd/kubeadm/app/componentconfigs/defaults.go

```diff
@@ -64,14 +67,24 @@ func DefaultKubeProxyConfiguration(internalcfg *kubeadmapi.ClusterConfiguration)
 }
 
 // DefaultKubeletConfiguration assigns default values for the kubelet ComponentConfig
-func DefaultKubeletConfiguration(internalcfg *kubeadmapi.ClusterConfiguration) {
+func DefaultKubeletConfiguration(internalcfg *kubeadmapi.ClusterConfiguration, nodeReg *kubeadmapi.NodeRegistrationOptions) {
 	externalkubeletcfg := &kubeletconfigv1beta1.KubeletConfiguration{}
 
 	// Do a roundtrip to the external version for defaulting
 	if internalcfg.ComponentConfigs.Kubelet != nil {
 		Scheme.Convert(internalcfg.ComponentConfigs.Kubelet, externalkubeletcfg, nil)
 	}
 
+	if nodeReg.CRISocket == constants.DefaultDockerCRISocket {
+		execer := utilsexec.New()
+		driver, err := kubeadmutil.GetCgroupDriverDocker(execer)
```

Comment:
this will not work in all setups, but also dynamically detecting the cgroup driver in DefaultKubeletConfiguration is not the right place for this.

we use the tool `kind` for e2e testing. kind uses `kubeadm config images list`:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/77059/pull-kubernetes-e2e-kind/1122122067366383616/build-log.txt
`kubeadm config images list` ends up calling `GetCgroupDriverDocker` but Docker is not running in the build container. we should avoid detecting a cgroup driver as part of the defaulting functions.

unfortunately the required change is much more complicated.
for kubeadm init the driver needs to be detected before the KubeletConfiguration file is written to disk:
https://github.com/kubernetes/kubernetes/blob/a404238d1c90ee00827624249b9223e7644e3f99/cmd/kubeadm/app/cmd/phases/init/kubelet.go#L71

for kubeadm join the KubeletConfiguration is downloaded from the cluster and written to disk:
https://github.com/kubernetes/kubernetes/blob/a404238d1c90ee00827624249b9223e7644e3f99/cmd/kubeadm/app/cmd/phases/join/kubelet.go#L134-L137
the KubelConfiguration that was downloaded needs to be modified with the detected cgroup driver before written to disk.

i'm starting to think that we should not do this refactor as is, but instead think how to translate all the flags from:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubelet/flags.go
to the KubeletConfiguration and not only `cgroupDriver`.

/hold

Reply:
Yes, the kubelet config content is generated too early, it is better to generate before writing it to disk, just like the `kubeadm-flags.env` file.